pkg/cert/truststore.go

```diff
@@ -72,6 +77,56 @@ func (m *fileTrustStore) AddCertificate(certificate *x509.Certificate) error {
 	return m.save()
 }
 
+// GetRoots checks if certificates have the same issuer as the subject and if they are self signed
+// multiple roots can be active at the same time.
+func (m *fileTrustStore) GetRoots(moment time.Time) []*x509.Certificate {
+	var certs []*x509.Certificate
+
+	for _, c := range m.certs {
+		if isSelfSigned(c, moment) {
+			certs = append(certs, c)
+		}
+	}
+
+	return certs
+}
+
+func (m *fileTrustStore) GetCertificates(chain [][]*x509.Certificate, moment time.Time, isCA bool) [][]*x509.Certificate {
+	var certs [][]*x509.Certificate
+	pool := x509.NewCertPool()
+
+	// construct pool with signers and its signers
+	for _, subChain := range chain {
+		for _, c := range subChain {
+			pool.AddCert(c)
+		}
+	}
+
+	for _, c := range m.certs {
+		if c.IsCA == isCA {
+			chain, err := c.Verify(x509.VerifyOptions{Roots: pool, CurrentTime: moment})
+			if err == nil {
+				for _, subChain := range chain {
+					certs = append(certs, subChain)
+				}
+			}
+		}
+	}
+
+	return certs
+}
+
+func isSelfSigned(cert *x509.Certificate, moment time.Time) bool {
+	if bytes.Equal(cert.RawIssuer, cert.RawSubject) && cert.IsCA {
+		pool := x509.NewCertPool()
+		pool.AddCert(cert)
+		chain, err := cert.Verify(x509.VerifyOptions{Roots: pool, CurrentTime: moment})
```

Comment:
I think this function does too much, given the name of the func `isSelfSigned`:
1. checks whether it's self-signed
2. checks whether it's valid on the given moment in time

When the certificate isn't valid at the given moment, it would seem the certificate isn't self-signed since `false` is returned.

I'd leave out the validity check and do that somewhere else.

Reply:
had to change it anyway, so now the validity check is done in `GetRoots`